### PR TITLE
Addition of arguments to ensemble.py and simulation.py for configuration writing. Minor addition file_manager.py

### PR DESCRIPTION
**Pull request by DaveCasson**
The focus of the changes is the addition of options to ensemble and simulation to allow specification of the frequency of restart, writing the configuration, to allow the whole configuration not to be written and for the file manager alone to be updated.

Changes are also made to avoid the automatic use of the run_suffix for naming the simulation object. This is problematic for daily timestepping, where the suffix is consistent between ensemble members.

file_manager is update to allow the statePath to be specified and used.

### CODE REVIEW

File: pysumma/ensemble.py [hunk 1]
```diff
@@ -154,9 +156,11 @@ def start(self, run_option: str='local', prerun_cmds: list=None):
             # Sleep calls are to ensure writeout happens
             config = self.configuration[n]
             self.submissions.append(self._client.submit(
-                _submit, s, n, run_option, prerun_cmds, config))
+                _submit, s, n, run_option, run_suffix, prerun_cmds, freq_restart,write_config,write_manager,config))
```
Comment (arbennett) [inline pysumma/ensemble.py]: Could you split this so that it doesn't end up being a long line?

File: pysumma/ensemble.py [hunk 2]
```diff
@@ -154,9 +156,11 @@ def start(self, run_option: str='local', prerun_cmds: list=None):
...
 
-    def run(self, run_option: str='local', prerun_cmds=None, monitor: bool=True):
+    def run(self, run_option: str='local', run_suffix: str='',
+              prerun_cmds: list=None, freq_restart: str=None, write_config: bool=True,
+              write_manager: bool=False, monitor: bool=True):
```
Comment (arbennett) [inline pysumma/ensemble.py]: Fix indentation

File: pysumma/ensemble.py [hunk 3]
```diff
@@ -169,7 +173,7 @@ def run(self, run_option: str='local', prerun_cmds=None, monitor: bool=True):
         monitor:
             Whether to halt operation until runs are complete
         """
-        self.start(run_option, prerun_cmds)
+        self.start(run_option, run_suffix, prerun_cmds, freq_restart, write_config, write_manager)
```
Comment (arbennett) [inline pysumma/ensemble.py]: Split into 2 lines

File: pysumma/ensemble.py [hunk 4]
```diff
@@ -194,8 +198,10 @@ def monitor(self):
         Halt computation until submitted simulations are complete
         """
         simulations = self._client.gather(self.submissions)
-        for s in simulations:
-            self.simulations[s.run_suffix] = s
+        names = [n for n,s in self.simulations.items()]
+
+        for s,name in zip(simulations,names):
+            self.simulations[name] = s
```
Comment (arbennett) [inline pysumma/ensemble.py]: I'm guessing this was some sort of a bug before, but could you explain why this change was made?

Also, a complete nitpick, but you can do:
```
names = self.simulations.values()
```
instead of
```
names = [n for n,s in self.simulations.items()]
```

File: pysumma/ensemble.py [hunk 5]
```diff
@@ -232,18 +239,19 @@ def rerun_failed(self, run_option: str='local', prerun_cmds=None,
             s = self.simulations[n]
             s.reset()
             self.submissions.append(self._client.submit(
-                _submit, s, n, run_option, prerun_cmds, config))
+                _submit, s, n, run_option, run_suffix, prerun_cmds, freq_restart,write_config,write_manager, config))
```
Comment (arbennett) [inline pysumma/ensemble.py]: Split into 2 lines

File: pysumma/ensemble.py [hunk 6]
```diff
@@ -232,18 +239,19 @@ def rerun_failed(self, run_option: str='local', prerun_cmds=None,
...
         if monitor:
             return self.monitor()
         else:
             return True
 
-
-def _submit(s: Simulation, name: str, run_option: str, prerun_cmds: List,
-            config: Dict, **kwargs):
+def _submit(s: Simulation, name: str, run_option: str,  run_suffix: str,
+              prerun_cmds: list, freq_restart: str,write_config: bool, write_manager: bool,config: Dict, **kwargs):
```
Comment (arbennett) [inline pysumma/ensemble.py]: Split into 2 lines

File: pysumma/simulation.py [hunk 7]
```diff
@@ -302,7 +304,7 @@ def _run_docker(self, run_suffix, processes=1,
 
     def start(self, run_option='local',  run_suffix='pysumma_run', processes=1,
               prerun_cmds=[], startGRU=None, countGRU=None, iHRU=None,
-              freq_restart=None, progress=None, **kwargs):
+              freq_restart=None, write_config=True,write_manager=False, progress=None, **kwargs):
```
Comment (arbennett) [inline pysumma/simulation.py]: Split into 2 lines

File: pysumma/simulation.py [hunk 8]
```diff
@@ -360,7 +365,7 @@ def run(self, run_option='local',  run_suffix='pysumma_run', processes=1,
             hourly output.
         """
         self.start(run_option, run_suffix, processes, prerun_cmds,
-                   startGRU, countGRU, iHRU, freq_restart, progress, **kwargs)
+                   startGRU, countGRU, iHRU, freq_restart,write_config, write_manager, progress, **kwargs)
```
Comment (arbennett) [inline pysumma/simulation.py]: Split into 2 lines

File: pysumma/simulation.py [hunk 9]
```diff
@@ -324,7 +329,7 @@ def start(self, run_option='local',  run_suffix='pysumma_run', processes=1,
 
     def run(self, run_option='local',  run_suffix='pysumma_run', processes=1,
             prerun_cmds=None, startGRU=None, countGRU=None, iHRU=None,
-            freq_restart=None, progress=None, **kwargs):
+            freq_restart=None, write_config=True,write_manager=False,progress=None, **kwargs):
```
Comment (arbennett) [inline pysumma/simulation.py]: Split into 2 lines

File: pysumma/simulation.py [hunk 10]
```diff
@@ -311,7 +313,10 @@ def start(self, run_option='local',  run_suffix='pysumma_run', processes=1,
         if not prerun_cmds:
             prerun_cmds = []
         self.run_suffix = run_suffix
-        self._write_configuration(name=run_suffix)
+        if write_config:
+            self._write_configuration(name=run_suffix)
+        if write_manager:
+            self._write_file_manager()
```
Comment (arbennett) [inline pysumma/simulation.py:320]: I'm wondering here if we want to have `_write_configuration` take an argument that's just `write_manager` and then roll `_write_file_manager` into `_write_configuration`, since it's only one line of code.

**Review by arbennett — CHANGES_REQUESTED** [review]
This looks great, just some minor formatting stuff. I haven't been _strictly_ following any code styleguides like PEP8 or black, but maybe it's getting to the point where we should...

File: pysumma/simulation.py (hunk 10 above)
Comment (DaveCasson) [reply pysumma/simulation.py:320]: I would suggest keep separate, as a clear writing of the manager only. This to me is a quite a separate action / update then writing the entire config.

File: pysumma/ensemble.py (hunk 4 above)
Comment (DaveCasson) [reply pysumma/ensemble.py]: Previously that the run_suffix was taken always as the  name. This was implemented to that the first key of the dictionary is used, allowing different ensemble naming. In my use case, allowing the same run_suffix to be used.

The simplified code that works is 
names = self.simulations.keys()

File: pysumma/ensemble.py (hunk 1 above)
Comment (DaveCasson) [reply pysumma/ensemble.py]: Done

File: pysumma/ensemble.py (hunk 2 above)
Comment (DaveCasson) [reply pysumma/ensemble.py]: Done

File: pysumma/ensemble.py (hunk 3 above)
Comment (DaveCasson) [reply pysumma/ensemble.py]: Done

File: pysumma/ensemble.py (hunk 5 above)
Comment (DaveCasson) [reply pysumma/ensemble.py]: Done

File: pysumma/ensemble.py (hunk 4 above)
Comment (arbennett) [reply pysumma/ensemble.py]: Ah, that makes sense - thanks!

File: pysumma/simulation.py (hunk 10 above)
Comment (arbennett) [reply pysumma/simulation.py:320]: Fair enough, sounds good to me.

**Review by arbennett — APPROVED** [review]
Looks good to me!
